File: mobile/ios/Alma/Screens/AlmaLaunch.swift
import SwiftUI

/// The first two and a half seconds: a star chart drawing itself, then folding
/// into the mark.
///
/// **What was here before.** `UILaunchScreen` names a colour, so iOS drew a
/// flat night rectangle and SwiftUI replaced it in one frame with a finished
/// screen. Nothing arrived; something was simply there instead of something
/// else. Against a backend on the same machine that state lasts a few dozen
/// milliseconds, which makes it worse rather than better — the app appears to
/// flicker. The first fix faded three things up over a second, and the owner's
/// verdict was the right one: not enough happening, and none of it about stars.
///
/// **What this is.** One clock, five overlapping movements, 2.8 seconds:
///
/// | from | to | what |
/// |---|---|---|
/// | 0.00 | 0.90 | the field of stars fades up and the chart settles in from 4% larger |
/// | 0.20 | 1.55 | six nodes light one after another, along the chart |
/// | 0.30 | 1.70 | the lines between them draw themselves, `trim` 0 → 1 |
/// | 1.50 | 2.30 | one comet crosses, and the chart contracts and dims |
/// | 1.70 | 2.55 | the mark blooms where the brightest node was, then the wordmark |
///
/// It overlaps on purpose. Five things each starting when the last has finished
/// is a slideshow; five whose curves cross is a single movement, and that
/// difference is most of what "expensive" means in motion.
///
/// **The mark does not spin.** `AlmaStar`'s own documentation forbids it —
/// *never in a circle, never rotated, never used as a loading indicator* — and
/// a rotating four-pointed sparkle is the exact cliché the mark was drawn to
/// avoid. It blooms in place, out of the chart it belongs to.
///
/// **Why it may take this long.** A launch state that appears for 40ms and
/// vanishes is a flash, which reads as a fault, and this is the one screen
/// every session begins with. When the session is genuinely slower the screen
/// simply stays, so the floor costs nothing on the path it was written for.
struct AlmaLaunch: View {

    /// Called once the arrival has finished *and* the caller says the app is
    /// ready. Both, not either: leaving early is the jump cut this replaces.
    var ready: Bool
    var done: () -> Void

    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    /// The whole arrival, and the floor on how long this screen is shown.
    /// Grew from 2.8 with the second draft: the owner's verdict on the first
    /// was "слабенькая", and the fix was not speed but density — a zodiac
    /// nine nodes instead of six, a second comet,
    /// and motes rising into the bloom all need the extra half-second.
    private static let runtime: Double = 3.4

    @State private var start = Date.timeIntervalSinceReferenceDate
    @State private var dwelt = false

    var body: some View {
        ZStack {
            Color.almaNight.ignoresSafeArea()

            // `NightSky` as a sibling rather than through `.nightSky()`: the
            // modifier puts the sky *behind* its content, and here the sky
            // arriving is the content. A background cannot be seen to fade.
            //
            // `.ceremony` rather than `.cabinet`, and not for brightness: the
            // cabinet mood draws its own comet, and two crossing at the same
            // moment reads as weather rather than as a sky. This screen draws
            // the only one.
            NightSky(mood: .ceremony, seed: 0x414C_4D41)
                .ignoresSafeArea()
                .opacity(reduceMotion ? 1 : eased(0, 0.9))

            if reduceMotion {
                // The same picture, finished, with nothing moving. Removing the
                // screen entirely would not be an accommodation — it would be
                // the flash this exists to prevent.
                chart(at: Self.runtime)
            } else {
                TimelineView(.animation) { timeline in
                    chart(at: timeline.date.timeIntervalSinceReferenceDate - start)
                }
            }
        }
        .task {
            try? await Task.sleep(for: .milliseconds(Int(Self.runtime * 1000)))
            dwelt = true
        }
        // **Both edges, and neither of them read from inside the task.**
        //
        // Deciding inside `.task` after the sleep hung the app. A `Task`
        // closure captures the view value it was created with, and this view is
        // created while `ready` is false; the session finished a tenth of a
        // second later, so `onChange` fired while `dwelt` was still false, and
        // the sleep then woke holding a `ready` that had been false when it was
        // captured and was never updated. Two halves each waiting for the
        // other, and a launch screen that never left — measured, still there
        // after four seconds against a local backend.
        //
        // `onChange` bodies run against a freshly-made view, so both of these
        // see both current values. Whichever lands second is the one that goes.
        .onChange(of: ready) { _, _ in handOver() }
        .onChange(of: dwelt) { _, _ in handOver() }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(Text(L10n.stateLoading))
    }

    private func handOver() {
        guard ready, dwelt else { return }
        done()
    }

    // MARK: — the drawing

    @ViewBuilder
    private func chart(at t: Double) -> some View {
        // How far the chart has folded away, and how far the mark has come up.
        // Derived from one clock so the two can never drift apart and leave a
        // frame holding both or neither.
        let fold = eased(1.90, 2.65, at: t)
        let bloom = eased(2.05, 2.85, at: t)
        let word = eased(2.35, 3.05, at: t)

        ZStack {
            StarChart(
                lines: eased(0.35, 1.95, at: t),
                nodes: eased(0.25, 1.80, at: t),
                comet: eased(1.20, 2.20, at: t),
                cometTwo: eased(1.65, 2.60, at: t),
                time: t
            )
            .frame(width: 300, height: 300)
            // Almost all the way out, not most of the way. At 0.88 the chart
            // was still legible behind the mark and the two read as clutter
            // rather than as one resolving into the other.
            .opacity(1 - fold * 0.97)
            // Contracting rather than only fading: the chart is being drawn
            // into the mark, and a shape that merely dims has been abandoned
            // rather than resolved.
            .scaleEffect(1.04 - eased(0, 0.9, at: t) * 0.04 - fold * 0.10)

            VStack(spacing: 24) {
                AlmaStar()
                    .frame(width: 46, height: 46)
                    .opacity(bloom)
                    .scaleEffect(0.72 + bloom * 0.28)
                    .background {
                        // Six motes rising into the bloom — sparks drawn up
                        // by the mark as it lights, gone once it has.
                        ZStack {
                            ForEach(0..<6, id: \.self) { i in
                                let phase = min(max(bloom * 1.4 - Double(i) * 0.12, 0), 1)
                                Circle()
                                    .fill(Color.almaGoldBright.opacity(0.7 * phase * (1 - phase) * 4 * 0.35))
                                    .frame(width: 2.5, height: 2.5)
                                    .offset(
                                        x: CGFloat([-26, 18, -12, 30, -34, 8][i]),
                                        y: CGFloat(34 - phase * 74)
                                    )
                            }
                        }
                    }

                Text(verbatim: "ALMA")
                    .font(AlmaFonts.ui(15))
                    // The letters settle: tracking closes from scattered to
                    // set as the word fades in — type arriving, not appearing.
                    .tracking(7 + (1 - word) * 9)
                    .foregroundStyle(Color.almaParchment.opacity(0.9))
                    .opacity(word)
                    .offset(y: (1 - word) * 7)

                // **The part that says "loading", and the part I dropped.**
                //
                // The first version of this screen ended on `AlmaPresence` —
                // Alma's own light, breathing — and the rewrite replaced it
                // with a chart and forgot to put it back. The owner's report
                // was that there is no loading, twice, and both times it was
                // this: a sequence that plays and stops is a title card, and a
                // title card in front of a person who is waiting does not tell
                // them anything is happening.
                //
                // `AlmaPresence` is the one loop the brand book allows, for
                // exactly this reason — stillness there reads as a hang. Under
                // it, a hairline that fills over the run: the quietest
                // possible progress, in one gold, no box.
                VStack(spacing: 16) {
                    AlmaPresence(size: 18, ring: false)
                        .opacity(word * 0.9)

                    Capsule()
                        .fill(Color.almaGold.opacity(0.16))
                        .frame(width: 84, height: 1)
                        .overlay(alignment: .leading) {
                            Capsule()
                                .fill(Color.almaGoldBright.opacity(0.75))
                                .frame(width: 84 * min(t / Self.runtime, 1), height: 1)
                        }
                        .opacity(eased(0.55, 1.20, at: t))
                }
                .padding(.top, 6)
            }
        }
    }

    /// A value that goes 0 → 1 between two moments, on an ease-out curve.
    ///
    /// Ease-out rather than linear everywhere: a light that arrives at a
    /// constant rate reads as a progress bar. The cubic is the cheapest curve
    /// that decelerates convincingly, and using one curve for every element is
    /// what makes five movements look like one hand.
    private func eased(_ from: Double, _ to: Double, at t: Double? = nil) -> Double {
        let now = t ?? (Date.timeIntervalSinceReferenceDate - start)
        let raw = min(max((now - from) / (to - from), 0), 1)
        return 1 - pow(1 - raw, 3)
    }
}

/// Nine stars, three chains of lines, and two comets — the sky connecting
/// itself, nothing drawn around it.
///
/// **Still not a real constellation, deliberately.** A recognisable Orion on
/// the launch screen is a claim about the sky tonight, and this screen has no
/// chart to make it from. It is the same visual language as the journey's
/// `ConstellationArt`, denser: the first draft read as a sketch, and the
/// owner said so. Density, not speed, is what changed.
private struct StarChart: View {

    /// 0 → 1 as the lines draw themselves.
    var lines: Double
    /// 0 → 1 as the nodes light, one after another.
    var nodes: Double
    /// The two comets. Off screen at both ends of their runs.
    var comet: Double
    var cometTwo: Double
    /// The raw clock, for the twinkle that never stops.
    var time: Double

    private let box: CGFloat = 300

    private let points: [(x: CGFloat, y: CGFloat, r: CGFloat)] = [
        (48, 196, 3.2), (104, 118, 4.6), (168, 152, 3.0),
        (214, 58, 5.2), (268, 104, 3.4), (132, 224, 3.8),
        (76, 64, 2.6), (232, 186, 3.0), (176, 250, 2.4),
    ]

    var body: some View {
        ZStack {
            // No dial and no ring — the owner's call: the arrival is the sky
            // connecting itself, nothing drawn around it.
            spine.trim(from: 0, to: lines)
                .stroke(Color.almaGold.opacity(0.5), lineWidth: 1)
            branch.trim(from: 0, to: max(0, lines * 1.3 - 0.3))
                .stroke(Color.almaGold.opacity(0.24), lineWidth: 1)
            branchTwo.trim(from: 0, to: max(0, lines * 1.5 - 0.5))
                .stroke(Color.almaGold.opacity(0.18), lineWidth: 1)

            ForEach(points.indices, id: \.self) { i in
                let share = Double(i) / Double(points.count)
                let lit = min(max((nodes - share * 0.55) / 0.45, 0), 1)
                // The twinkle: each star breathes on its own beat once lit,
                // so the finished chart is alive rather than printed.
                let breathe = 0.75 + 0.25 * sin(time * (1.1 + Double(i % 4) * 0.4) + Double(i) * 2)
                let p = points[i]
                Circle()
                    .fill(Color.almaStarFill)
                    .frame(width: p.r * 2, height: p.r * 2)
                    .shadow(color: Color.almaGoldBright.opacity(0.55 * breathe), radius: 6)
                    .scaleEffect((0.4 + lit * 0.6) * breathe)
                    .opacity(lit)
                    .position(x: p.x, y: p.y)
            }

            // The ring around the brightest node — where the mark blooms from.
            Circle()
                .stroke(Color.almaGoldBright.opacity(0.22 * nodes), lineWidth: 1)
                .frame(width: 30 + (1 - nodes) * 14, height: 30 + (1 - nodes) * 14)
                .position(x: 214, y: 58)

            cometView(comet, width: 90, angle: 28,
                      from: CGPoint(x: -40, y: 250), to: CGPoint(x: 340, y: 40))
            cometView(cometTwo, width: 64, angle: -16,
                      from: CGPoint(x: 330, y: 210), to: CGPoint(x: -30, y: 120))
        }
        .frame(width: box, height: box)
        .compositingGroup()
    }

    @ViewBuilder
    private func cometView(_ phase: Double, width: CGFloat, angle: Double,
                           from: CGPoint, to: CGPoint) -> some View {
        if phase > 0 && phase < 1 {
            Capsule()
                .fill(
                    LinearGradient(
                        colors: [.clear, Color.almaStarFill.opacity(0.75), .clear],
                        startPoint: .leading, endPoint: .trailing
                    )
                )
                .frame(width: width, height: 1.2)
                .rotationEffect(.degrees(angle))
                .position(
                    x: from.x + (to.x - from.x) * phase,
                    y: from.y + (to.y - from.y) * phase
                )
                .opacity(sin(phase * .pi))
        }
    }

    private var spine: Path {
        Path { p in
            p.move(to: CGPoint(x: 48, y: 196))
            p.addLine(to: CGPoint(x: 104, y: 118))
            p.addLine(to: CGPoint(x: 168, y: 152))
            p.addLine(to: CGPoint(x: 214, y: 58))
            p.addLine(to: CGPoint(x: 268, y: 104))
        }
    }

    private var branch: Path {
        Path { p in
            p.move(to: CGPoint(x: 104, y: 118))
            p.addLine(to: CGPoint(x: 132, y: 224))
            p.addLine(to: CGPoint(x: 214, y: 58))
        }
    }

    private var branchTwo: Path {
        Path { p in
            p.move(to: CGPoint(x: 76, y: 64))
            p.addLine(to: CGPoint(x: 104, y: 118))
            p.move(to: CGPoint(x: 232, y: 186))
            p.addLine(to: CGPoint(x: 268, y: 104))
            p.move(to: CGPoint(x: 132, y: 224))
            p.addLine(to: CGPoint(x: 176, y: 250))
        }
    }
}

File: mobile/android/app/src/main/kotlin/ai/pazl/alma/ui/components/AlmaLaunch.kt
package ai.pazl.alma.ui.components

import ai.pazl.alma.R
import ai.pazl.alma.ui.sky.NightSky
import ai.pazl.alma.ui.sky.SkyConfig
import ai.pazl.alma.ui.theme.AlmaFonts
import ai.pazl.alma.ui.theme.AlmaPalette
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.produceState
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.scale
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.PathMeasure
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.rotate
import androidx.compose.ui.graphics.drawscope.scale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.em
import androidx.compose.ui.unit.sp
import androidx.compose.ui.util.lerp
import android.provider.Settings
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.sin
import kotlinx.coroutines.android.awaitFrame

/**
 * The first two and a half seconds: a star chart drawing itself, then folding
 * into the mark. The iOS `AlmaLaunch`, on Compose.
 *
 * One clock, five overlapping movements, 2.8 seconds: the field of stars fades
 * up, six nodes light one after another, the lines between them draw
 * themselves, one comet crosses as the chart contracts, and the mark blooms
 * where the brightest node was — then the wordmark, Alma's own light, and the
 * quietest possible progress hairline.
 *
 * It overlaps on purpose. Five things each starting when the last has finished
 * is a slideshow; five whose curves cross is a single movement, and that
 * difference is most of what "expensive" means in motion.
 *
 * **The mark does not spin.** `AlmaStar`'s own documentation forbids it, and a
 * rotating four-pointed sparkle is the exact cliché the mark was drawn to
 * avoid. It blooms in place, out of the chart it belongs to.
 *
 * [done] is called once the arrival has finished *and* [ready] is true. Both,
 * not either: leaving early is the jump cut this replaces, and when the session
 * is genuinely slower the screen simply stays.
 */
@Composable
fun AlmaLaunch(ready: Boolean, done: () -> Unit) {
    val context = LocalContext.current
    // The platform's "remove animations" accessibility setting zeroes the
    // animator scale. The same picture, finished, with nothing moving —
    // removing the screen entirely would not be an accommodation, it would be
    // the flash this exists to prevent.
    val still = remember {
        Settings.Global.getFloat(
            context.contentResolver, Settings.Global.ANIMATOR_DURATION_SCALE, 1f
        ) == 0f
    }

    // The one clock every movement reads. Stops advancing shortly after the
    // run so a slow backend does not keep a whole screen invalidating at 120Hz
    // for nothing — `AlmaPresence` breathes on its own animation.
    val t by produceState(if (still) Runtime else 0f) {
        if (still) return@produceState
        val start = awaitFrame()
        while (value < Runtime + 0.1f) {
            val now = awaitFrame()
            value = ((now - start) / 1e9).toFloat()
        }
    }

    val dwelt = t >= Runtime
    LaunchedEffect(dwelt, ready) {
        if (dwelt && ready) done()
    }

    val loading = stringResource(R.string.state_loading)
    NightSky(config = SkyConfig(seed = 0x414C, motes = 0, comet = false)) {
        Box(
            Modifier
                .fillMaxSize()
                .alpha(if (still) 1f else eased(0f, 0.9f, t))
                .clearAndSetSemantics { contentDescription = loading },
            contentAlignment = Alignment.Center,
        ) {
            // How far the chart has folded away, and how far the mark has come
            // up — derived from the one clock so the two can never drift apart.
            val fold = eased(1.55f, 2.20f, t)
            val bloom = eased(1.70f, 2.45f, t)
            val word = eased(2.00f, 2.55f, t)

            StarChartCanvas(
                lines = eased(0.30f, 1.70f, t),
                nodes = eased(0.20f, 1.55f, t),
                comet = eased(1.50f, 2.30f, t),
                modifier = Modifier
                    .size(300.dp)
                    // Almost all the way out, not most of the way: at less the
                    // chart stays legible behind the mark and the two read as
                    // clutter rather than as one resolving into the other.
                    .alpha(1f - fold * 0.97f)
                    .scale(1.04f - eased(0f, 0.9f, t) * 0.04f - fold * 0.10f),
            )

            Column(horizontalAlignment = Alignment.CenterHorizontally) {
                Box(Modifier.alpha(bloom).scale(0.72f + bloom * 0.28f)) {
                    AlmaStar(size = 46.dp)
                }

                Spacer(Modifier.height(24.dp))
                Text(
                    text = "ALMA",
                    style = TextStyle(
                        fontFamily = AlmaFonts.Sans,
                        fontSize = 15.sp,
                        letterSpacing = 0.45.em,
                        color = AlmaPalette.Parchment.copy(alpha = 0.9f),
                    ),
                    modifier = Modifier.alpha(word),
                )

                // The part that says "loading": Alma's own light, breathing —
                // the one loop the brand allows, because stillness here reads
                // as a hang — and under it a hairline that fills over the run.
                Spacer(Modifier.height(24.dp))
                Box(Modifier.alpha(word * 0.9f)) {
                    AlmaPresence(size = 18.dp, ring = false)
                }
                Spacer(Modifier.height(16.dp))
                Canvas(Modifier.size(width = 84.dp, height = 1.dp).alpha(eased(0.55f, 1.20f, t))) {
                    drawLine(
                        color = AlmaPalette.Gold.copy(alpha = 0.16f),
                        start = Offset(0f, 0f),
                        end = Offset(size.width, 0f),
                        strokeWidth = size.height,
                        cap = StrokeCap.Round,
                    )
                    drawLine(
                        color = AlmaPalette.GoldBright.copy(alpha = 0.75f),
                        start = Offset(0f, 0f),
                        end = Offset(size.width * min(t / Runtime, 1f), 0f),
                        strokeWidth = size.height,
                        cap = StrokeCap.Round,
                    )
                }
            }
        }
    }
}

/** The whole arrival, and the floor on how long this screen is shown. */
private const val Runtime = 2.8f

/**
 * A value that goes 0 → 1 between two moments, on an ease-out cubic.
 *
 * Ease-out rather than linear everywhere: a light that arrives at a constant
 * rate reads as a progress bar. One curve for every element is what makes five
 * movements look like one hand.
 */
private fun eased(from: Float, to: Float, at: Float): Float {
    val raw = ((at - from) / (to - from)).coerceIn(0f, 1f)
    return 1f - (1f - raw).pow(3)
}

/**
 * Six stars, four lines between them, and one comet.
 *
 * **Not a real constellation, deliberately.** A recognisable Orion on the
 * launch screen is a claim about the sky tonight, and this screen has no chart
 * to make it from — the whole product is built on not saying things it has not
 * calculated.
 */
@Composable
private fun StarChartCanvas(
    lines: Float,
    nodes: Float,
    comet: Float,
    modifier: Modifier = Modifier,
) {
    // On a 300-point grid, so the coordinates read as a drawing.
    val points = remember {
        listOf(
            Offset(48f, 196f) to 3.2f, Offset(104f, 118f) to 4.6f,
            Offset(168f, 152f) to 3.0f, Offset(214f, 58f) to 5.2f,
            Offset(268f, 104f) to 3.4f, Offset(132f, 224f) to 3.8f,
        )
    }
    val spine = remember {
        Path().apply {
            moveTo(48f, 196f); lineTo(104f, 118f); lineTo(168f, 152f)
            lineTo(214f, 58f); lineTo(268f, 104f)
        }
    }
    val branch = remember {
        Path().apply {
            moveTo(104f, 118f); lineTo(132f, 224f); lineTo(214f, 58f)
        }
    }
    val spineMeasure = remember { PathMeasure().apply { setPath(spine, false) } }
    val branchMeasure = remember { PathMeasure().apply { setPath(branch, false) } }
    val trimmed = remember { Path() }

    Canvas(modifier) {
        val unit = size.minDimension / 300f
        fun at(p: Offset) = Offset(p.x * unit, p.y * unit)

        fun draw(measure: PathMeasure, share: Float, color: Color) {
            if (share <= 0f) return
            trimmed.reset()
            measure.getSegment(0f, measure.length * share, trimmed, true)
            scale(unit, unit, pivot = Offset.Zero) {
                drawPath(trimmed, color, style = Stroke(width = 1f / unit))
            }
        }
        draw(spineMeasure, lines, AlmaPalette.Gold.copy(alpha = 0.5f))
        // A quarter-run behind the spine, so the branch is still being drawn
        // when the spine arrives at its last node.
        draw(branchMeasure, (lines * 1.3f - 0.3f).coerceIn(0f, 1f), AlmaPalette.Gold.copy(alpha = 0.24f))

        points.forEachIndexed { index, (p, r) ->
            // Each node opens in its own slice of the run, so they light
            // *along* the chart rather than together.
            val share = index / points.size.toFloat()
            val lit = ((nodes - share * 0.55f) / 0.45f).coerceIn(0f, 1f)
            if (lit <= 0f) return@forEachIndexed
            val centre = at(p)
            val radius = r * unit * (0.4f + lit * 0.6f)
            drawCircle(
                brush = Brush.radialGradient(
                    colors = listOf(
                        AlmaPalette.GoldBright.copy(alpha = 0.5f * lit),
                        Color.Transparent,
                    ),
                    center = centre,
                    radius = radius * 3.2f,
                ),
                radius = radius * 3.2f,
                center = centre,
            )
            drawCircle(
                color = AlmaPalette.StarFill.copy(alpha = lit),
                radius = radius,
                center = centre,
            )
        }

        // The ring around the brightest node — the place the mark blooms from,
        // marked before it is used, and closing as the chart lights.
        drawCircle(
            color = AlmaPalette.GoldBright.copy(alpha = 0.22f * nodes),
            radius = (15f + (1f - nodes) * 7f) * unit,
            center = at(Offset(214f, 58f)),
            style = Stroke(width = 1f),
        )

        if (comet > 0f && comet < 1f) {
            // Brightest at the middle of its crossing rather than at the
            // edges, so it is never seen to appear or to stop.
            val head = at(
                Offset(
                    lerp(-40f, 340f, comet),
                    lerp(250f, 40f, comet),
                )
            )
            rotate(degrees = 28f, pivot = head) {
                drawLine(
                    brush = Brush.horizontalGradient(
                        colors = listOf(
                            Color.Transparent,
                            AlmaPalette.StarFill.copy(alpha = 0.75f * sin(comet * Math.PI).toFloat()),
                            Color.Transparent,
                        ),
                        startX = head.x - 45f * unit,
                        endX = head.x + 45f * unit,
                    ),
                    start = Offset(head.x - 45f * unit, head.y),
                    end = Offset(head.x + 45f * unit, head.y),
                    strokeWidth = 1.2f * unit,
                    cap = StrokeCap.Round,
                )
            }
        }
    }
}
